src/steps/resource-manager/index.ts

```diff
@@ -393,6 +403,40 @@ export async function fetchResourceManagerProject(
   await jobState.addEntity(projectEntity);
 }
 
+export async function fetchIamPolicyAuditConfig(
+  context: IntegrationStepContext,
+): Promise<void> {
+  const {
+    instance: { config },
+    jobState,
+  } = context;
+  const client = new ResourceManagerClient({ config });
+
+  await client.iteratePolicyAuditConfigs(async (auditConfig) => {
+    const auditConfigEntity = createAuditConfigEntity(auditConfig);
+    if (auditConfigEntity) {
+      await jobState.addEntity(auditConfigEntity);
+
+      const serviceEntity = await jobState.findEntity(
+        getServiceApiEntityKey({
+          projectId: client.projectId,
+          serviceApiName: auditConfig.service as string,
+        }),
+      );
+
+      if (serviceEntity) {
+        await jobState.addRelationship(
+          createDirectRelationship({
+            _class: RelationshipClass.MONITORS,
+            from: auditConfigEntity,
+            to: serviceEntity,
+          }),
+        );
+      }
```

Comment:
I like this audit_config -MONITORS-> service relationship!
I'm a bit worried about the following:

> Specifies a service that will be enabled for audit logging. For example, `storage.googleapis.com`, `cloudsql.googleapis.com`. `allServices` is a special value that covers all services.
My guess is if the value is `allServices` that `getServiceApiEntityKey()` won't work - it can't, it returns a single service meanwhile we'd want to create multiple relationships.

We'd want to modify the code so in the case where the value is `allServices` we somehow get a list of all the services (I'm not sure if we have functionality in place to do this right now, or if we need to add some `jobState` util function) and then create relationships to each.

Furthermore - `Schema$AuditLogConfig` has `exemptedMembers?: string[] | null;`
We may want to try and include some of that data when we're building relationships between audit_config and service perhaps? Maybe data on the edge?

**Here's a bit more info from the typedefs:**
The configuration determines which permission types are logged, and what identities, if any, are exempted from logging.

An AuditConfig must have one or more AuditLogConfigs.

If there are AuditConfigs for both `allServices` and a specific service, the union of the two AuditConfigs is used for that service: the log_types specified in each AuditConfig are enabled, and the exempted_members in each AuditLogConfig are exempted.

Example Policy with multiple AuditConfigs: 
```
{
  "audit_configs": [
    {
      "service": "allServices",
      "audit_log_configs": [{
          "log_type": "DATA_READ",
          "exempted_members": [ "user:jose@example.com" ] 
        }, {
          "log_type": "DATA_WRITE" 
        }, {
          "log_type": "ADMIN_READ" 
        }
      ]
    }, {
      "service": "sampleservice.googleapis.com",
      "audit_log_configs": [{
          "log_type": "DATA_READ"
        }, { 
          "log_type": "DATA_WRITE",
          "exempted_members": [ "user:aliya@example.com" ] 
        }
      ] 
    }
  ]
}
```

For sampleservice, this policy enables DATA_READ, DATA_WRITE and ADMIN_READ logging. It also exempts jose@example.com from DATA_READ logging, and aliya@example.com from DATA_WRITE logging.

Interested in hearing Austin's thoughts on the above!

Reply:
Hmmm, I think I can handle `allServices` using `jobState.iterateEntities`.

I couldn't ingest the `exempted_members` because I can't think of a way to preserve the connection between `exempted_members` to `log_type` since we can only ingest primitives/array or primitives. I might need some ideas here.

The possible approach that I'm thinking for this is maybe ingesting the `log_type` as a separate entity and just build a `HAS` relationship between the `auditConfig` entity and the `log_type` entity. What do you think?

If we figure out how to do the above topic, I can also build a relationship with the exempted members since we also ingest `google_user`. Something like `google_audit_config_restricts_user`.

Reply:
Thanks for sharing your ideas! :+1: I'll ask a few people and see what are they thinking about all of this.

Reply:
@eXtremeX @RPGPH 

The `exemptedMembers` is super interesting! Nice find! If I'm understanding correctly, adding a member to the `exemptedMembers` would mean that their specific action would _not_ be audited? Someone could add themselves to an `exemptedMembers` list and then perform all sorts of actions while flying under the radar! 🤔

It's not immediately clear from the docs (https://cloud.google.com/resource-manager/reference/rest/Shared.Types/AuditConfig#auditlogconfig) whether `exemptedMembers` shares the same type of format as IAM resource bindings...IAM bindings can have the following formats (see: https://cloud.google.com/iam/docs/reference/rest/v1/Policy):

- `user:mike@example.com`
- `group:admins@example.com`
- `domain:google.com`
- `serviceAccount:my-project-id@appspot.gserviceaccount.com`

There is code in the repo that takes an IAM member as input and parses the data into a consistent format. We use that to build the following:

```
google_iam_binding -ASSIGNED-> google_user
google_iam_binding -ASSIGNED-> google_domain
google_iam_binding -ASSIGNED-> google_group
google_iam_binding -ASSIGNED-> google_iam_service_account
```

See: https://github.com/JupiterOne/graph-google-cloud/blob/ce17b9db52f798d571b2037c38a2b1131fa901af/src/steps/resource-manager/index.ts#L76

----

I do think we'd probably want to land on a _mapped_ relationship between the `google_cloud_audit_config` and the member, but attaching this data as a property on the relationship makes a lot of sense too!

If we think in terms of a J1 query, perhaps we'd ask a question like:

> Which individual Google Cloud members have service API audit exemptions?

```
FIND google_cloud_audit_config
THAT ALLOWS (google_user | google_group | google_domain | google_iam_service_account)
```

The `ALLOWS` relationship could have a property on it that represents the `log_type`:

> Which individual Google Cloud members have service API audit exemptions that allow write access?

```
FIND google_cloud_audit_config
THAT ALLOWS AS r (google_user | google_group | google_domain | google_iam_service_account)
WHERE r.logType ~= "WRITE"
```

Reply:
The `exemptedMembers` feature will be addressed on a separate PR, I will use this discussion as a reference. Thanks!